rts/lra/jaxrs/src/main/java/io/narayana/lra/filter/ServerLRAFilter.java

```diff
@@ -460,46 +575,149 @@ private boolean isJaxRsCancel(ContainerRequestContext requestContext, ContainerR
+    // the following structure is used to track progress so that such failures can be reported in the response
+    // filter processing
+    private enum ProgressStep {
+        Left ("leave succeeded"),
+        LeaveFailed("leave failed"),
+        Started("start ok"),
+        StartFailed("start failed"),
+        Joined("join succeeded"),
+        JoinFailed("join failed"),
+        Ended("end ok"),
+        CloseFailed("close failed"),
+        CancelFailed("cancel failed");
+
+        final String status;
+
+        ProgressStep(final String status) {
+            this.status = status;
+        }
 
-        return lraClient.startLRA(parentLRA, clientId, timeout, ChronoUnit.MILLIS);
+        @Override
+        public String toString() {
+            return status;
+        }
     }
 
-    private void resumeTransaction(URI lraId) {
-        // nothing to do
+    // list of steps (both successful and unsuccesful) performed so far by the request and response filter
+    // and is used for error reporting
+    private static class Progress {
+        static EnumSet<ProgressStep> failures = EnumSet.of(
+                ProgressStep.LeaveFailed,
+                ProgressStep.StartFailed,
+                ProgressStep.JoinFailed,
+                ProgressStep.CloseFailed,
+                ProgressStep.CancelFailed);
+
+        ProgressStep progress;
+        String reason;
+
+        public Progress(ProgressStep progress, String reason) {
+            this.progress = progress;
+            this.reason = reason;
+        }
+
+        public boolean ok() {
+            return !failures.contains(progress);
+        }
     }
 
-    private String getCompensatorId(URI lraId, UriInfo uriInfo, Long timeout) {
-        Map<String, String> terminateURIs = NarayanaLRAClient.getTerminationUris(resourceInfo.getResourceClass(), uriInfo, timeout);
+    // convert the list of steps carried out by the filters into a warning message
+    private String processLRAOperationFailures(ArrayList<Progress> progress) {
+        StringJoiner badOps = new StringJoiner(", ");
+        StringJoiner goodOps = new StringJoiner(", ");
+        StringBuffer code = new StringBuffer("-");
 
-        if (!terminateURIs.containsKey("Link")) {
-            throwGenericLRAException(lraId, Response.Status.BAD_REQUEST.getStatusCode(),
-                    "Missing complete or compensate annotations");
-        }
+        progress.forEach(p -> {
+            if (p.ok()) {
+                code.insert(0, p.progress.ordinal());
+                goodOps.add(String.format("%s (%s)", p.progress.name(), p.progress.status));
+            } else {
+                code.append(p.progress.ordinal());
+                badOps.add(String.format("%s (%s)", p.progress.name(), p.reason));
+            }
+        });
+
+        /*
+         * return a string which encodes the result:
+         * <major code>-<failed op codes>-<successful op codes>: <details of failed ops> (<details of successful ops>)
```

Comment:
After a while of decoding I understand the idea here.
I would use probably a format but it's a matter of taste as discussed elsewhere.
I would rather ask here what happens when there are multiple failed (or successful) op codes. They seem to be appended without any delimiter, or not?

In fact I wonder what is the purpose of having these opcodes as part of the user readable failure message?

Reply:
Not. The code is using `java.util.StringJoiner` which includes the delimiter in the constructor.
The purpose is for knowing which operations are potentially in doubt.
We need some response to tell the caller that something is in doubt and this gives him a fighting chance of figuring out what went wrong i.e. it's not meant for users to read, rather it is for application logic to reason about the status of the system.

You/we can provide something more comprehensive in MP release 1.x

Reply:
We need to make progress and we don't have time for long protracted coding style debates.
After we have finally got the code in front of users we can take our foot off the pedal and you can rework it all to satisfy your sensibilities.